Sources/PackageGraph/PackageGraphLoader.swift

```diff
@@ -222,9 +243,21 @@ private func createResolvedPackages(
         let package = packageBuilder.package
 
         // Establish the manifest-declared package dependencies.
-        packageBuilder.dependencies = package.manifest.package.dependencies.compactMap({
+        let dependencies = package.manifest.package.dependencies.compactMap({
             packageMap[PackageReference.computeIdentity(packageURL: $0.url)]
         })
+        
+        // Detect duplicate dependencies
+        for duplicate in dependencies.findDuplicates() {
+            let diagnostic = DuplicateDependencyDiagnostic(
+                dependencyName: duplicate.package.name,
+                version: duplicate.package.manifest.version
+            )
+            diagnostics.emit(diagnostic)
+        }
+        
+        // Remove duplicate dependencies
+        packageBuilder.dependencies = Array(Set(dependencies))
```

Comment:
This will be expensive, we should enhance findDuplicates to return a tuple with unique and duplicate elements.